### PR TITLE
Enable GitLab CI on Tioga and Dane

### PR DESCRIPTION
**Pull request by balos1**

### CODE REVIEW

File: scripts/spack/repo.yaml
Comment (gardner48) [inline scripts/spack/repo.yaml:1]: What is this file needed for? Does it need to be here or can it also go under `.gitlab`?

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -10,127 +16,95 @@
...
-  script:
-    - echo ${ALLOC_NAME}
-    - bsub -q ${BENCHMARK_QUEUE} -J ${ALLOC_NAME} -nnodes ${BENCHMARK_NNODES} -W ${DEFAULT_TIME} -Is .gitlab/build_and_bench.sh
+# Template for jobs triggering a build-and-test sub-pipeline:
+.build-and-test:
+  stage: build-and-test
+  trigger:
+    include:
+      - local: '.gitlab/custom-jobs-and-variables.yml'
+      - project: 'sundials/radiuss-shared-ci' # https://lc.llnl.gov/gitlab/sundials/radiuss-shared-ci
```
Comment (gardner48) [inline .gitlab-ci.yml:87]: In the docs it would be helpful to clarify that `.build-and-test` pulls in files from a mirror of our fork of `radiuss-shared-ci`.

Comment (gardner48) [reply .gitlab-ci.yml:87]: Also it looks like you had to add dane to the radiuss-shared-ci repo, will this be pushed back upstream?

**Review by gardner48 — CHANGES_REQUESTED** [review]

File: .gitlab/uberenv
[the remark this answers is not in the document]
Comment (balos1) [reply .gitlab/uberenv:1]: They were enhancements, not fixes. A [PR](https://github.com/LLNL/uberenv/pulls) for one of the features is open already, and I will open another for the other feature as well.

File: .gitlab/radiuss-spack-configs
[the remark this answers is not in the document]
Comment (balos1) [reply .gitlab/radiuss-spack-configs:1]: It was enhancements rather than fixes, but I will try to upstream them (if they are acceptable for radiuss more broadly).

File: doc/superbuild/source/developers/testing/CI.rst [hunk 2]
```diff
@@ -111,6 +111,11 @@ For information specific to the LLNL GitLab CI see:
 * `LLNL GitLab Runner Tags <https://lc.llnl.gov/gitlab/public-info/gitlab-ci/-/wikis/Gitlab-CI-Basic-Information>`_
 
 
+SUNDIALS utilizes the GitLab CI pipeline code repository shared by LLNL RADIUSS
+projects. The `docs <https://radiuss-shared-ci.readthedocs.io/en/latest/>`__ for
```
[the remark this answers is not in the document]
Comment (balos1) [reply doc/superbuild/source/developers/testing/CI.rst:115]: Steven and I ran into an issue with conflicting references recently and that is when I discovered this syntax.

File: scripts/spack/packages/sundials/package.py
[the remark this answers is not in the document]
Comment (balos1) [reply scripts/spack/packages/sundials/package.py:1]: I would like to have it here as I would like us to merge the upstream Spack package.py and this one, then keep this copy up-to-date continuously. I.e., use it for more than just gitlab.

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -10,127 +16,95 @@
...
 
+# We define the following GitLab pipeline variables:
 variables:
+##### LC GITLAB CONFIGURATION
+# Use an LLNL service user to run CI. This prevents from running pipelines as
+# an actual user.
+  LLNL_SERVICE_USER: ""
+# Use the service user workspace. Solves permission issues, stores everything
+# at the same location whoever triggers a pipeline.
+#  CUSTOM_CI_BUILDS_DIR: ""
```
[the remark this answers is not in the document]
Comment (balos1) [reply .gitlab-ci.yml:47]: I just symlink the home directory location to a place in `/usr/workspace/` instead. I dont think we need to configure this until we setup a service user.

File: .gitlab-ci.yml [hunk 4]
```diff
@@ -10,127 +16,95 @@
...
 
-stages:
-  - q_build_and_test
-  - l_build_and_test
-  - l_build_and_bench
-  - c_build_and_test
+##### PROJECT VARIABLES
+# We build the projects in the CI clone directory.
+# Used in script/gitlab/build_and_test.sh script.
+# TODO: add a clean-up mechanism.
```
[the remark this answers is not in the document]
Comment (balos1) [reply .gitlab-ci.yml:54]: I would rather not as then it will make diffing with the template from radiuss-shared-ci harder.

File: .gitlab/custom-jobs-and-variables.yml [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+# Arguments for top level allocation
+  DANE_SHARED_ALLOC: "--exclusive --partition=pdebug --time=60 --nodes=1"
+# Arguments for job level allocation
+  DANE_JOB_ALLOC: "--time=55 --nodes=1 --overlap"
+# Add variables that should apply to all the jobs on a machine:
+#  DANE_MY_VAR: "..."
+
+# Tioga
+# Arguments for top level allocation
+  TIOGA_SHARED_ALLOC: "--exclusive --time-limit=60m --nodes=1"
```
[the remark this answers is not in the document]
Comment (balos1) [reply .gitlab/custom-jobs-and-variables.yml:25]: I think this is getting added elsewhere already but I will double check.

File: .gitlab/jobs/tioga.yml [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+#   error while loading shared libraries: libpgmath.so: cannot open shared object file: No such file or directory
+# This seems like it might be an LC environment issue, so disabling the job for now.
+tioga_rocmcc_571_tpls:
+  parallel:
+    matrix:
+      - COMPILER_SPEC: rocmcc@5.7.1
+        AMDGPU_TARGET: [gfx90a]
+  variables:
+    ON_TIOGA: "OFF" # disable until we can figure out libpgmath.so error
+    # have to use ginkgo@master because our spack version does not have ginkgo@1.8.0: yet (which seems to be needed)
```
[the remark this answers is not in the document]
Comment (balos1) [reply .gitlab/jobs/tioga.yml:51]: Our interface does not, but I could not get ginkgo older than 1.8.0 to compile with rocm@5.7.1 or 6.2.0.

File: scripts/spack/repo.yaml
Comment (balos1) [reply scripts/spack/repo.yaml:1]: `scripts/packages` is technically a spack repo which our CI picks up. This file should go here because it is part of the repo. As I said elsewhere, it is best to have the spack package.py for sundials in `scripts/` as it is not gitlab specific and will soon be used as our primary place to update our spack package (updated will be copied to upstream).

File: .gitlab-ci.yml (hunk 1 above)
Comment (balos1) [reply .gitlab-ci.yml:87]: The PR to upstream it is already open [here](https://github.com/LLNL/radiuss-shared-ci/pull/38).

File: .gitlab-ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (balos1) [reply .gitlab-ci.yml:47]: I see the radiuss-shared-ci docs recommend setting it if not using a service user. I guess we can set it to `/usr/workspace/sundials/ci/.builds/`.

File: cmake/macros/SundialsAddTest.cmake [hunk 7]
```diff
@@ -186,11 +186,11 @@ macro(SUNDIALS_ADD_TEST NAME EXECUTABLE)
                                              OR (SUNDIALS_TEST_MPIRUN_COMMAND)))
         if(SUNDIALS_TEST_MPIRUN_COMMAND)
           set(RUN_COMMAND
-              "${SUNDIALS_TEST_MPIRUN_COMMAND} ${MPIEXEC_NUMPROC_FLAG} ${SUNDIALS_ADD_TEST_MPI_NPROCS} ${MPIEXEC_PREFLAGS}"
+              "${SUNDIALS_TEST_MPIRUN_COMMAND} ${MPIEXEC_PREFLAGS} ${MPIEXEC_NUMPROC_FLAG} ${SUNDIALS_ADD_TEST_MPI_NPROCS} ${MPIEXEC_POSTFLAGS}"
```
[the remark this answers is not in the document]
Comment (balos1) [reply cmake/macros/SundialsAddTest.cmake:189]: The [current version of the CMake documentation is clear](https://cmake.org/cmake/help/latest/module/FindMPI.html#usage-of-mpiexec).  

>${MPIEXEC_EXECUTABLE} ${MPIEXEC_NUMPROC_FLAG} ${MPIEXEC_MAX_NUMPROCS}
  ${MPIEXEC_PREFLAGS} EXECUTABLE ${MPIEXEC_POSTFLAGS} ARGS

I am going to update our usage to match that.

[the remark this answers is not in the document]
Comment (balos1) [reply cmake/macros/SundialsAddTest.cmake:189]: Ah, yes, putting preflags before the numprocs argument does not work with `flux` because the PREFLAGS for flux has to be set to `run` since the command is `flux run` (and CMake will error our if you set `MPIEXEC_EXECUTABLE to `flux run`).

[the remark this answers is not in the document]
Comment (gardner48) [reply cmake/macros/SundialsAddTest.cmake]: Ah, I was only looking at the variable descriptions and just needed to scroll down. On another project we do the following in our run script and it seems to work on Tioga 
```
mpiexec_executable=${MPIEXEC_EXECUTABLE:-"srun"}
# If using flux, append "run" after the flux executable path
if [[ "${mpiexec_executable}" == "flux" ]]
then
    mpiexec_executable="$(which ${mpiexec_executable}) run"
    flux jobs
    flux resource list
else
    mpiexec_executable="$(which ${mpiexec_executable})"
fi
```
where `MPIEXEC_EXECUTABLE` is an environment variable set by the job then in the `cmake` line we have
```
-DMPIEXEC_EXECUTABLE="${mpiexec_executable}" \
```
```

File: scripts/spack/packages/sundials/package.py
[the remark this answers is not in the document]
Comment (gardner48) [reply scripts/spack/packages/sundials/package.py:1]: Sounds good, should add this as a task in the backlog.

File: .gitlab-ci.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (gardner48) [reply .gitlab-ci.yml:47]: I think you have to manually create the symlink which is why they recommend setting it explicitly. Having this be in the shared workspace makes sense to me.

File: .gitlab/jobs/tioga.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (gardner48) [reply .gitlab/jobs/tioga.yml:51]: Ok, just wanted to make sure this was a spack/ginkgo issue and not that a new version dependence had snuck in some now.

File: cmake/macros/SundialsAddTest.cmake (hunk 7 above)
[the remark this answers is not in the document]
Comment (balos1) [reply cmake/macros/SundialsAddTest.cmake]: OK I went with the solution from your other project.

[the remark this answers is not in the document]
Comment (balos1) [reply cmake/macros/SundialsAddTest.cmake]: Oops, I was looking at the wrong job. The other project approach is causing a failure.

File: scripts/spack/repo.yaml
Comment (gardner48) [reply scripts/spack/repo.yaml:1]: Should we have a `spack` directory then for all things spack related?

Comment (balos1) [reply scripts/spack/repo.yaml:1]: Do you mean a top-level `spack` directory instead of `scripts/spack`?

Comment (gardner48) [reply scripts/spack/repo.yaml:1]: That's what I was thinking since I don't really think of this as a script but maybe that's making the top level too cluttered.

Comment (balos1) [reply scripts/spack/repo.yaml:1]: I think it would make the top level to cluttered and "script" is pretty generic, so it should OK.

File: doc/shared/sundials/Install.rst [hunk 8]
```diff
@@ -868,6 +868,23 @@ illustration only.
...
+   Specifies flags that come before the MPI executable, but after ``MPIEXEC_EXECUTABLE``,
+   ``MPIEXEC_NUMPROC_FLAG``, and ``MPIEXEC_MAX_NUMPROCS``.
+
+   Default: none
+
+   .. note:: This option is triggered only if MPI is enabled (``ENABLE_MPI`` is ``ON``).
+
+.. cmakeoption:: MPIEXEC_POSTFLAGS
+
+   Specifies flags that come after the MPI executable but before any other program arguments.
```
Comment (gardner48) [inline doc/shared/sundials/Install.rst:882]: Minor rewording to avoid confusing the "MPI executable" with `mpiexec`
```suggestion
   Specifies flags that come after the executable to run but before any other program arguments.
```

File: doc/superbuild/source/developers/testing/CI.rst [hunk 9]
```diff
@@ -123,125 +128,103 @@ the jobs in a stage succeed, the pipeline moves on to the next stage. If any job
...
+      matrix:
+        - COMPILER_SPEC: rocmcc@6.2.0
+          AMDGPU_TARGET: [gfx90a]
+    variables:
+      # have to use ginkgo@master because our spack version does not have ginkgo@1.8.0: yet (which seems to be needed)
+      # similarly, we need a newer magma than available to compile with 'rocm@6:' so we turn it off
+      SPEC: "%${COMPILER_SPEC} cstd=99 cxxstd=14 precision=double amdgpu_target=${AMDGPU_TARGET} +rocm+mpi~magma+ginkgo+kokkos ^ginkgo@master+rocm amdgpu_target=${AMDGPU_TARGET} ^kokkos+rocm amdgpu_target=${AMDGPU_TARGET}"
+    before_script:
+      - module load rocmcc/6.2.0-magic
+    extends: [.sundials_job_on_tioga]
```
Comment (gardner48) [inline doc/superbuild/source/developers/testing/CI.rst:180]: Indent three space to align with `code-block::`
```suggestion
   tioga_rocmcc_620_tpls:
     parallel:
       matrix:
         - COMPILER_SPEC: rocmcc@6.2.0
           AMDGPU_TARGET: [gfx90a]
     variables:
       # have to use ginkgo@master because our spack version does not have ginkgo@1.8.0: yet (which seems to be needed)
       # similarly, we need a newer magma than available to compile with 'rocm@6:' so we turn it off
       SPEC: "%${COMPILER_SPEC} cstd=99 cxxstd=14 precision=double amdgpu_target=${AMDGPU_TARGET} +rocm+mpi~magma+ginkgo+kokkos ^ginkgo@master+rocm amdgpu_target=${AMDGPU_TARGET} ^kokkos+rocm amdgpu_target=${AMDGPU_TARGET}"
     before_script:
       - module load rocmcc/6.2.0-magic
     extends: [.sundials_job_on_tioga]
```

File: doc/superbuild/source/developers/testing/CI.rst [hunk 10]
```diff
@@ -123,125 +128,103 @@ the jobs in a stage succeed, the pipeline moves on to the next stage. If any job
...
+    parallel:
+      matrix:
+        - COMPILER_SPEC: rocmcc@6.2.0
+          AMDGPU_TARGET: [gfx90a]
+    variables:
+      ON_TIOGA: "OFF" # disable this job
+      SPEC: "%${COMPILER_SPEC} cstd=99 cxxstd=14 precision=double amdgpu_target=${AMDGPU_TARGET} +rocm+mpi~magma+ginkgo+kokkos ^ginkgo@master+rocm amdgpu_target=${AMDGPU_TARGET} ^kokkos+rocm amdgpu_target=${AMDGPU_TARGET}"
+    before_script:
+      - module load rocmcc/6.2.0-magic
+    extends: [.sundials_job_on_tioga]
```
Comment (gardner48) [inline doc/superbuild/source/developers/testing/CI.rst:209]: Indent three spaces to align with `code-block::`
```suggestion
   tioga_rocmcc_620_tpls:
     parallel:
       matrix:
         - COMPILER_SPEC: rocmcc@6.2.0
           AMDGPU_TARGET: [gfx90a]
     variables:
       ON_TIOGA: "OFF" # disable this job
       SPEC: "%${COMPILER_SPEC} cstd=99 cxxstd=14 precision=double amdgpu_target=${AMDGPU_TARGET} +rocm+mpi~magma+ginkgo+kokkos ^ginkgo@master+rocm amdgpu_target=${AMDGPU_TARGET} ^kokkos+rocm amdgpu_target=${AMDGPU_TARGET}"
     before_script:
       - module load rocmcc/6.2.0-magic
     extends: [.sundials_job_on_tioga]
```

**Review by gardner48 — CHANGES_REQUESTED** [review]
A couple of minor things, otherwise this looks good

**Review by gardner48 — APPROVED** [review]

File: .gitlab/build_and_test.sh [hunk 11]
```diff
@@ -88,7 +88,6 @@ then
         mkdir "${buildcache}"
     fi
 
-    chmod g+rwx -R $buildcache
```
Comment (gardner48) [inline .gitlab/build_and_test.sh]: Should we note in the dev docs that this needs to be done manually (and similarly for other directories under the shared `ci` directory)?

**Review by gardner48 — APPROVED** [review]
